src/node_errors.cc

```diff
@@ -661,44 +595,39 @@ const char* errno_string(int errorno) {
   }
 }
 
-}  // namespace errors
-
-void DecorateErrorStack(Environment* env,
-                        const errors::TryCatchScope& try_catch) {
-  Local<Value> exception = try_catch.Exception();
-
-  if (!exception->IsObject()) return;
-
-  Local<Object> err_obj = exception.As<Object>();
-
-  if (IsExceptionDecorated(env, err_obj)) return;
-
-  AppendExceptionLine(env, exception, try_catch.Message(), CONTEXTIFY_ERROR);
-  TryCatchScope try_catch_scope(env);  // Ignore exceptions below.
-  MaybeLocal<Value> stack = err_obj->Get(env->context(), env->stack_string());
-  MaybeLocal<Value> maybe_value =
-      err_obj->GetPrivate(env->context(), env->arrow_message_private_symbol());
-
-  Local<Value> arrow;
-  if (!(maybe_value.ToLocal(&arrow) && arrow->IsString())) {
-    return;
-  }
+MaybeLocal<Value> PrepareStackTraceCallback(Local<Context> context,
+                                            Local<Value> exception,
+                                            Local<Array> trace) {
+  Environment* env = Environment::GetCurrent(context);
+  Local<Function> prepare = env->prepare_stack_trace_callback();
+  CHECK(!prepare.IsEmpty());
+  Local<Value> args[] = {
+      context->Global(),
+      exception,
+      trace,
+      errors::CreateMessage(env, exception.As<Object>()),
+  };
+  return prepare->Call(context, Undefined(env->isolate()), 4, args);
+}
 
-  if (stack.IsEmpty() || !stack.ToLocalChecked()->IsString()) {
-    return;
-  }
+void SetPrepareStackTraceCallback(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args[0]->IsFunction());
+  env->set_prepare_stack_trace_callback(args[0].As<Function>());
+  env->isolate()->SetPrepareStackTraceCallback(PrepareStackTraceCallback);
```

Comment:
We usually do this in `NewIsolate()`, and in the callback, return if `env->prepare_stack_trace_callback()` is empty.

Reply:
the return value of the callback is used to create the stack property. if env or the callback are missing for some reason, what's the sane thing to do?

Reply:
The API here seems inadequate, it should be capable of allowing the caller to make it a noop.
It’s preferred not to call per-Isolate setters from within JS during bootstrap because this would get lost in the snapshot and we have to call it again after deserialization, so we might as well set it in C++ upfront along with other setters and reinstall them all together, instead of having these scattered around.

Reply:
I think when env is `nullptr`, you can just return `exception->ToString(context)`? That's similar to a noop.